proposals/0118-partitioned-epoch-reward-distribution.md

```diff
@@ -0,0 +1,273 @@
+formula to round up to the nearest integer without using floating point value
+arithmetic:
+
+```
+M = ((4096 - 1)+num_stake_accounts)/4096
+```
+
+To safeguard against the number of stake accounts growing dramatically and
+overflowing the number of blocks in an epoch, the number of blocks is capped at
+10% of the total slots in an epoch. If the block cap is reached, accounts per
+partition is allowed to surpass the 4,096 target.
+
+The [SipHash 1-3](https://www.aumasson.jp/siphash/siphash.pdf) pseudo-random
+function is used to hash stake account addresses efficiently and uniformly
+across the blocks in the reward distribution phase. The hashing function for an
+epoch is created by seeding a new `SipHasher` with the parent block's blockhash.
+This hashing function can then be used to hash each active stake account's
+address into a u64 hash value. The reward distribution block index `I` can then
+be computed by applying the following formula to the hash:
+
+```
+I = (M * stake_address_hash) / 2^64
+```
+
+### Reward Distribution
+
+The reward distribution phase happens after the reward computation phase, ie.
+during the second block in the epoch as per the current design. It lasts for `M`
+blocks (see previous section).
+
+#### `EpochRewards` Sysvar Account
+
+The `EpochRewards` sysvar account records whether the rewards distribution phase
+is in progress, as well as the details needed to resume distribution when
+starting from a snapshot during the reward distribution phase. These details
+include: the parameters needed to recalculate the reward partitions, the total
+rewards to be distributed, and the amount of rewards distributed so far.
+
+The layout of `EpochRewards` sysvar is shown in the following pseudo code.
+Prefixes indidate the offset of each field, and padding bytes are always zero.
+
+```
+#[repr(C)] // C representation, 8-byte alignment
+struct EpochRewards{
+   // whether the rewards period (calculation and distribution) is active
+   // byte. false: 0x00, true: 0x01
+   /* 0x00 */ active: bool,
+
+   // little-endian unsigned 64-bit integer
+   /* 0x08 */ distribution_starting_block_height: u64,
+   
+   // little-endian unsigned 64-bit integer
+   /* 0x10 */ num_partitions: u64,
+   
+   // byte-array of length 32
+   /* 0x18 */ parent_blockhash: Hash,
+
+   // total points calculated for the current epoch, where points equals the sum
+   // of delegated stake * credits observed for all delegations
+   // little-endian unsigned 128-bit integer
+   /* 0x38 */ total_points: u128, 
+
+   // total rewards for the current epoch, in lamports
+   // little-endian unsigned 64-bit integer
+   /* 0x48 */ total_rewards: u64,
+
+   // distributed rewards for the current epoch, in lamports
+   // little-endian unsigned 64-bit integer
+   /* 0x50 */ distributed_rewards: u64, 
+}
+```
+
+The `EpochRewards` sysvar is repopulated at the start of the first block of the
```

Comment:
We need to add detail on when the sysvar is updated during the calculation phase and the distribution phase. That is to say, the reward calculation and sysvar update therein should probably be updated before the transactions in the block (or after, i dont think it matters, actually after is better). The rewards distribution phase and sysvar update should be done _after_ the transactions like rent collection.

Reply:
Okay before is fine. I think I could have been clearer that it is preferable to keep it consistent.